### PR TITLE
More detailed error messages on deser fail and timeout

### PR DESCRIPTION
**Pull request by QuantumEntangledAndy**
While I was making changes to the deserialiser I noticed that any errors during derserialisation would be silently ignored. Instead it would just drop the connection and restart.

This makes changes so that it prints the error prior to dropping and restarting the connection.

I also noticed that the error `std::sync::mpsc::RecvTimeoutError` is actually an enum of, `Timeout` or `Disconnected` but in the code we treat them the same as `Timeout` (this is why sometimes it would say timeout error even when the the 5 seconds (of `RX_TIMEOUT`) had yet to pass)

This also makes changes to the `Error` enum so that we correctly distinguish between the two.

### CODE REVIEW

File: src/bc_protocol/connection.rs [hunk 1]
```diff
@@ -55,7 +55,15 @@ impl BcConnection {
...
-            while BcConnection::poll(&mut context, &conn, &mut subs).is_ok() {}
+            while {
+                match BcConnection::poll(&mut context, &conn, &mut subs) {
+                    Ok(_) => true,
+                    Err(err) => {
+                        error!("Deserialization error: {:?}", err);
+                        false
+                    }
+                }
+            } {}
```
Comment (thirtythreeforty) [inline src/bc_protocol/connection.rs:66]: I see what you're doing here, but the "naked" `while` is weird imho.  How about:
```
    let mut result;
    while { result = foo(); result.is_ok() } {}
    error!("Deserialization error: {:?}", result.unwrap_err);
    result.unwrap_err();
```

**Review by thirtythreeforty — CHANGES_REQUESTED** [review]
Small quibble, see above :)

File: src/bc_protocol/connection.rs (hunk 1 above)
Comment (QuantumEntangledAndy) [reply src/bc_protocol/connection.rs:66]: I like being verbose but alright I'll change it haha

**Comment by QuantumEntangledAndy** [comment]
Edited to less naked loop

File: src/bc_protocol/connection.rs (hunk 1 above)
Comment (QuantumEntangledAndy) [reply src/bc_protocol/connection.rs:62]: Done

**Comment by QuantumEntangledAndy** [comment]
I would also like to get a full backtrace with the error. Would you be OK with swapping from `err-derive` to something that supports backtraces?

**Comment by QuantumEntangledAndy** [comment]
Ugh backtraces are nightly only... why rust why...

**Review by thirtythreeforty — APPROVED** [review]

**Pull request merged**